### PR TITLE
Make docker image status more granular with the percentage done

### PR DESCRIPTION
**Pull request by adiprerepa**
### Reasons for making this change

Set status of docker image more granular.

### Related issues

#3174 

### Screenshots

image

### Checklist

* [x] I've added a screenshot of the changes, if this is a frontend change
* [x] I've added and/or updated tests, if this is a backend change
* [ ] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: codalab/worker/worker_run_state.py [hunk 1]
```diff
@@ -290,8 +291,9 @@ def mount_dependency(dependency, shared_file_system):
         if not dependencies_ready:
             status_message = status_messages.pop()
             if status_messages:
-                status_message += "(and downloading %d other dependencies and docker images)" % len(
-                    status_messages
+                status_message += (
+                    "\n(and downloading %d other dependencies and docker images)"
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: Do we need this newline?

File: codalab/worker/docker_image_manager.py [hunk 2]
```diff
@@ -243,10 +243,14 @@ def image_availability_state(image_spec, success_message, failure_message):
                 def download():
                     logger.debug('Downloading Docker image %s', image_spec)
                     try:
-                        self._docker.images.pull(image_spec)
+                        # client = docker.APIClient(base_url='unix://var/run/docker.sock')
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]: Can you remove the comments?

**Review by epicfaace — COMMENTED** [review]
Were you able to test this? Can you show a screenshot of how the status looks like now?

File: codalab/worker/worker_run_state.py (hunk 1 above)
Comment (adiprerepa) [reply codalab/worker/worker_run_state.py]: I think it looks weird without it.

**Comment by adiprerepa** [comment]
image

I used percentages, this is how it looks.

File: codalab/worker/worker_run_state.py (hunk 1 above)
Comment (epicfaace) [reply codalab/worker/worker_run_state.py]: Ok -- the screenshot on this PR seems to not have the newline -- can you change the screenshot to the latest one so I can see how it looks like with the newline added?

image

File: codalab/worker/docker_image_manager.py [hunk 3]
```diff
@@ -243,10 +243,13 @@ def image_availability_state(image_spec, success_message, failure_message):
                 def download():
                     logger.debug('Downloading Docker image %s', image_spec)
                     try:
-                        self._docker.images.pull(image_spec)
+                        client = docker.APIClient(base_url='unix://var/run/docker.sock')
+                        self.generator = client.pull(image_spec, stream=True, decode=True)
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]: Why do we need to create a new client as opposed to just running `self._docker.images.pull`?

File: codalab/worker/worker_run_state.py [hunk 4]
```diff
@@ -276,22 +277,32 @@ def mount_dependency(dependency, shared_file_system):
         docker_image = run_state.resources.docker_image
         image_state = self.docker_image_manager.get(docker_image)
         if image_state.stage == DependencyStage.DOWNLOADING:
-            status_messages.append(
-                'Pulling docker image: ' + (image_state.message or docker_image or "")
-            )
+            msg = 'Pulling docker image %s' % docker_image
+            if isinstance(image_state.message, dict) and image_state.message['progressDetail']:
+                msg += '(%d%%)' % (
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: I think it would be cleaner to instead just make sure that `image_state.message` is always a string (it's good practice in general not to have the same variable have multiple possible types in Python). This means that this code here would probably be moved over to `docker_image_manager.py` and you'd set `self._downloading[image_spec]['status']` equal to `"Pulling docker image (XX%)"`

File: codalab/worker/worker_run_state.py [hunk 5]
```diff
@@ -276,22 +277,32 @@ def mount_dependency(dependency, shared_file_system):
...
+
+            status_messages.append(msg)
             dependencies_ready = False
         elif image_state.stage == DependencyStage.FAILED:
             # Failed to pull image; -> CLEANING_UP
             message = 'Failed to download Docker image: %s' % image_state.message
             logger.error(message)
             return run_state._replace(stage=RunStage.CLEANING_UP, failure_message=message)
 
+        logger.debug(dependencies_ready)
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: rm this line

**Comment by epicfaace** [comment]
> image
> 
> I used percentages, this is how it looks.

Can you change this so that there's a space before the parenthesis `(` as well?

File: codalab/worker/docker_image_manager.py (hunk 3 above)
Comment (adiprerepa) [reply codalab/worker/docker_image_manager.py]: When I do this, the `generator` returned is not an iterator. I am not sure why this is the case - but it only works when you have a new `APIClient`

Comment (epicfaace) [reply codalab/worker/docker_image_manager.py]: Ok. What exactly is returned from the generator coming from `self._docker.images.pull`? Is `self._docker.images` also an `APIClient`?

Comment (adiprerepa) [reply codalab/worker/docker_image_manager.py]: its an `ImageCollection` I believe - I will have to double check.

Comment (adiprerepa) [reply codalab/worker/docker_image_manager.py]: It looks like when we use `self._docker.images.pull`, it calls the same `pull` as we do with the new `APIClient`, but omits the output.

Reference: https://github.com/docker/docker-py/blob/master/docker/models/images.py#L449

**Comment by adiprerepa** [comment]
image

This is what it looks like when there are dependencies. Imo it looks good

File: codalab/worker/docker_image_manager.py (hunk 3 above)
Comment (epicfaace) [reply codalab/worker/docker_image_manager.py]: Ah, I see. In that case, can we just replace this with `client.client.api.pull` then? This way, we don't have to instantiate a new `APIClient`.

File: codalab/worker/docker_image_manager.py [hunk 6]
```diff
@@ -243,10 +243,21 @@ def image_availability_state(image_spec, success_message, failure_message):
...
+                        for line in self.generator:
+                            self._downloading[image_spec]['status'] = (
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]: I think we can pull this line out of the for loop

File: codalab/worker/worker_run_state.py [hunk 7]
```diff
@@ -290,8 +289,9 @@ def mount_dependency(dependency, shared_file_system):
         if not dependencies_ready:
             status_message = status_messages.pop()
             if status_messages:
-                status_message += "(and downloading %d other dependencies and docker images)" % len(
-                    status_messages
+                status_message += (
+                    "\n(and downloading %d other dependencies and docker images)"
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: I don't see a newline in the latest screenshot you sent. I'm guessing perhaps the newline just doesn't show up in the HTML; in that case, let's just remove the newline.

File: codalab/worker/worker.py [hunk 8]
```diff
@@ -79,6 +79,7 @@ def __init__(
         # A flag indicating if the worker will exit if it encounters an exception
         exit_on_exception=False,  # type: bool
     ):
+
```
Comment (epicfaace) [inline codalab/worker/worker.py]: undo this change

File: codalab/worker/worker_run_state.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 from codalab.worker.fsm import DependencyStage, StateTransitioner
 from codalab.worker.worker_thread import ThreadDict
 
+
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: undo this change

File: codalab/worker/worker_run_state.py [hunk 10]
```diff
@@ -276,9 +277,7 @@ def mount_dependency(dependency, shared_file_system):
         docker_image = run_state.resources.docker_image
         image_state = self.docker_image_manager.get(docker_image)
         if image_state.stage == DependencyStage.DOWNLOADING:
-            status_messages.append(
-                'Pulling docker image: ' + (image_state.message or docker_image or "")
-            )
+            status_messages.append(image_state.message)
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]: I think we should keep the "Pulling docker image [image_spec]" here (and remove it from docker_image_manage.py), just for consistency with the rest of the code.

File: codalab/worker/docker_image_manager.py (hunk 6 above)
Comment (adiprerepa) [reply codalab/worker/docker_image_manager.py]: how do you mean?

Comment (epicfaace) [reply codalab/worker/docker_image_manager.py]: Oh, I see what you were doing. Never mind my comment

File: codalab/worker/worker_run_state.py (hunk 10 above)
Comment (adiprerepa) [reply codalab/worker/worker_run_state.py]: I see - but keep the percent stuff in the message in `docker_image_manager.py`

**Comment by adiprerepa** [comment]
> Ah, I see. In that case, can we just replace this with client.client.api.pull then? This way, we don't have to instantiate a new APIClient.

@epicfaace what do you mean? `client` is the field being instantiated

**Comment by adiprerepa** [comment]
I just used `self._docker.api.pull`

**Comment by epicfaace** [comment]
> I just used `self._docker.api.pull`

Oh, ok, that makes sense then, thanks for updating

File: codalab/worker/docker_image_manager.py [hunk 11]
```diff
@@ -243,10 +243,18 @@ def image_availability_state(image_spec, success_message, failure_message):
                 def download():
                     logger.debug('Downloading Docker image %s', image_spec)
                     try:
-                        self._docker.images.pull(image_spec)
+                        self.generator = self._docker.api.pull(image_spec, stream=True, decode=True)
+                        for line in self.generator:
+                            self._downloading[image_spec]['status'] = ''
+                            if line['status'] == 'Downloading':
+                                self._downloading[image_spec]['status'] = ' (%d%%)' % (
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]: Let's move the space to worker_run_state.py instead
```suggestion
                                self._downloading[image_spec]['status'] = '(%d%%)' % (
```

File: codalab/worker/worker_run_state.py [hunk 12]
```diff
@@ -277,7 +277,7 @@ def mount_dependency(dependency, shared_file_system):
         image_state = self.docker_image_manager.get(docker_image)
         if image_state.stage == DependencyStage.DOWNLOADING:
             status_messages.append(
-                'Pulling docker image: ' + (image_state.message or docker_image or "")
+                'Pulling docker image %s%s' % (docker_image, image_state.message)
```
Comment (epicfaace) [inline codalab/worker/worker_run_state.py]:
```suggestion
                'Pulling docker image %s %s' % (docker_image, image_state.message)
```

File: codalab/worker/docker_image_manager.py [hunk 13]
```diff
@@ -243,10 +243,18 @@ def image_availability_state(image_spec, success_message, failure_message):
                 def download():
                     logger.debug('Downloading Docker image %s', image_spec)
                     try:
-                        self._docker.images.pull(image_spec)
+                        self.generator = self._docker.api.pull(image_spec, stream=True, decode=True)
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]: No need to assign to an instance variable
```suggestion
```

File: codalab/worker/docker_image_manager.py [hunk 14]
```diff
@@ -243,10 +243,18 @@ def image_availability_state(image_spec, success_message, failure_message):
...
+                        for line in self.generator:
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py]:
```suggestion
                        for line in self._docker.api.pull(image_spec, stream=True, decode=True):
```

File: codalab/worker/docker_image_manager.py [hunk 15]
```diff
@@ -243,10 +243,18 @@ def image_availability_state(image_spec, success_message, failure_message):
...
+                            self._downloading[image_spec]['status'] = ''
+                            if line['status'] == 'Downloading':
```
Comment (epicfaace) [inline codalab/worker/docker_image_manager.py:249]:
```suggestion
                            # Set the status to the percent completed
                            if line['status'] == 'Downloading':
```

**Review by epicfaace — APPROVED** [review]
Looks good! Just a few more changes and then we'll be all set to merge.

**Comment by adiprerepa** [comment]
@epicfaace should be good to go!
